File: 1639-number-of-ways-to-form-a-target-string-given-a-dictionary/1639-number-of-ways-to-form-a-target-string-given-a-dictionary.py
class Solution:
    def numWays(self, words: List[str], target: str) -> int:
        MOD = 10 ** 9 + 7
        m, n = len(words[0]), len(target)
        charAtIndexCnt = [[0] * m for _ in range(128)]
        for word in words:
            for i, c in enumerate(word):
                charAtIndexCnt[ord(c)][i] += 1

        @lru_cache(None)
        def dp(k, i):
            if i == n:
                return 1
            if k == m:
                return 0
            c = target[i]
            ans = dp(k + 1, i)
            if charAtIndexCnt[ord(c)][k] > 0:
                ans += dp(k + 1, i + 1) * charAtIndexCnt[ord(c)][k]
                ans %= MOD
            return ans

        return dp(0, 0)

File: 1639-number-of-ways-to-form-a-target-string-given-a-dictionary/1639-number-of-ways-to-form-a-target-string-given-a-dictionary.cpp
class Solution:
    def numWays(self, words: List[str], target: str) -> int:
        MOD = 10**9 + 7
        m, n = len(words[0]), len(target)
        
        freq = [Counter() for _ in range(m)]
        
        for word in words:
            for i, char in enumerate(word):
                freq[i][char] += 1
        
        @lru_cache(None)
        def dp(i, j):
            if j == n:
                return 1
            if i == m:
                return 0
            return (dp(i + 1, j) + dp(i + 1, j + 1) * freq[i][target[j]]) % MOD
        
        return dp(0, 0)